### PR TITLE
[refactor] Redesign job scheduler backends

### LINKED ISSUE #74: Consider using the delegation pattern in the Job scheduler internal interface
And perhaps in other places, too.

Internal issue: https://madra.cscs.ch/scs/reframe/issues/489

### PR DESCRIPTION
**Pull request by vkarak**
Scheduler logic is now separated from the job information. The `Job` class still offers the same API to its clients, but it delegates all the scheduler-specific work to a scheduler backend. The advantage of this design is that we can easily and cleanly support custom configurations for the backend schedulers. All scheduler backends inherit from the `JobScheduler` base abstract class. This class is intentionally left without any implementation, because it needs to serve as a pure interface, which the scheduler backends must implement. The key difference of this interface compared to the current one is that these function take the user job as an argument instead referring to it through `self`.

I fixed also a minor bug in the local launcher, that actually didn't ignore `options`, if these were set after its creation (now this is also the only way to set the launcher options, which is how it is used in practice).

I have also enabled flexible allocation for the local scheduler, since it's trivial. It's always the current host. :-)

@ZQyou these changes will most probably break your PBS local scheduler adaptations, but the changes should be straightforward to adapt.

Fixes #74. 

- ~~I still need to check if the documentation shows up correctly.~~
- ~~I still need to fix the unit tests in CSCS CI.~~

### CODE REVIEW

File: reframe/core/schedulers/__init__.py [hunk 1]
```diff
@@ -222,75 +256,40 @@ def guess_num_tasks(self):
...
+        available_nodes = self.scheduler.allnodes()
         getlogger().debug('flex_alloc_nodes: total available nodes %s ' %
                           len(available_nodes))
 
         # Try to guess the number of tasks now
-        available_nodes = self.filter_nodes(
-            available_nodes, self.sched_access + self.options)
-
+        available_nodes = self.scheduler.filternodes(self, available_nodes)
+        print(available_nodes)
```
Comment (ekouts) [inline reframe/core/schedulers/__init__.py]: This print should probably leave.

File: unittests/test_launchers.py [hunk 2]
```diff
@@ -70,19 +76,24 @@ def expected_command(self):
...
+        emitted_command = self.run_command(self.minimal_job)
         self.assertEqual(self.expected_minimal_command, emitted_command)
 
 
 class TestSrunLauncher(_TestLauncher, unittest.TestCase):
     @property
     def launcher(self):
-        return getlauncher('srun')(options=['--foo'])
+        ret = getlauncher('srun')()
+        print('ret =', ret)
```
Comment (ekouts) [inline unittests/test_launchers.py]: This print should be removed as well.

**Comment by vkarak** [comment]
@ekouts @teojgo @victorusu You can now review. All unit tests should be passing now. I have still the wip tag, because I need to check the documentation.

File: reframe/core/schedulers/__init__.py [hunk 3]
```diff
@@ -14,12 +13,49 @@
...
+    @abc.abstractmethod
+    def cancel(self, job):
+        pass
+
+    @abc.abstractmethod
+    def finished(self, job):
+        pass
+
+
+class Job:
```
Comment (victorusu) [inline reframe/core/schedulers/__init__.py]: I really liked the design. The separation of `Job`  into `JobScheduler` and `Job` allows a simpler implementation of the different schedulers and their functionalities. Now, `Job` is a struct that holds info about the job state, which is populated by the scheduler that schedules it.

**Pull request merged**